File: benchmarks-QuantumOptics.jl/multiplication_sparse_dense.jl
using QuantumOptics
using BenchmarkTools
include("benchmarkutils.jl")

srand(0)

basename = "multiplication_sparse_dense"

samples = 2
evals = 5
cutoffs = [50:50:1000;]
S = [0.1, 0.01, 0.001]
Nrand = 5

function setup(N, s)
    b = GenericBasis(N)
    op1 = SparseOperator(b, sprand(Complex128, N, N, s))
    op2 = randoperator(b)
    result = DenseOperator(b)
    op1, op2, result
end

function f(op1, op2, result)
    operators.gemm!(Complex128(1., 0.), op1, op2, Complex128(0., 0.), result)
end

for s in S
    name = basename * "_" * replace(string(s), ".", "")
    println("Benchmarking: ", name)
    print("Cutoff: ")
    results = []
    for N in cutoffs
        print(N, " ")
        T = 0.
        for i=1:Nrand
            op1, op2, result = setup(N, s)
            T += @belapsed f($op1, $op2, $result) samples=samples evals=evals
        end
        push!(results, Dict("N"=>N, "t"=>T/Nrand))
    end
    println()
    benchmarkutils.save(name, results)
end

File: benchmarks-QuTiP/multiplication_sparse_dense.py
import scipy.sparse as sp
import benchmarkutils

basename = "multiplication_sparse_dense"

samples = 2
evals = 5
cutoffs = range(50, 1001, 50)
S = [0.1, 0.01, 0.001]
Nrand = 5

def setup(N, s):
    op1 = sp.rand(N, N, s, dtype=float)*0.2j
    op2 = sp.rand(N, N, 1., dtype=float)*0.1j
    return op1, op2

def f(op1, op2):
    return op1*op2

for s in S:
    name = basename + "_" + str(s).replace(".", "")
    print("Benchmarking:", name)
    print("Cutoff: ", end="", flush=True)
    results = []
    for N in cutoffs:
        print(N, "", end="", flush=True)
        T = 0.
        for i in range(Nrand):
            op1, op2 = setup(N, s)
            T += benchmarkutils.run_benchmark(f, op1, op2, samples=samples, evals=evals)
        results.append({"N": N, "t": T/Nrand})
    print()
    benchmarkutils.save(name, results)
